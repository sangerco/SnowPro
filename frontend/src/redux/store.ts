import { Action, compose } from "redux";
import { ThunkAction, configureStore } from "@reduxjs/toolkit";
import { logger } from "redux-logger";
import rootReducer from "./reducers/rootReducer";

const composeEnhancers =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(logger, composeEnhancers),
});
