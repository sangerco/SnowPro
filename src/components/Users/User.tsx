import React from "react";

interface UserProps {
  user: {
    id: string;
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    bio?: string;
    avatar?: string;
    isAdmin: boolean;
  };
}

const User: React.FC<UserProps> = ({ user }) => {
  return (
    <li key={user.id}>
      <p>{user.id}</p>
      <p>{user.username}</p>
      <p>{user.firstName}</p>
      <p>{user.lastName}</p>
      <p>{user.email}</p>
      <p>{user.bio}</p>
      <p>{user.avatar}</p>
      <p>{user.isAdmin}</p>
    </li>
  );
};

export default User;
