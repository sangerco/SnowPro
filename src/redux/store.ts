import { configureStore, combineReducers } from "@reduxjs/toolkit";
import userReducer from "./slices/userSlice";
import authReducer from "./slices/authSlice";
import messageReducer from "./slices/messageSlice";
import messageReplyReducer from "./slices/messageReplySlice";
import reviewReducer from "./slices/reviewSlice";
import reviewReplyReducer from "./slices/reviewReplySlice";
import tagReducer from "./slices/tagSlice";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["messages", "messageReplies", "reviews", "reviewReplies"],
};

const rootReducer = combineReducers({
  users: userReducer,
  auth: authReducer,
  messages: messageReducer,
  messageReplies: messageReplyReducer,
  reviews: reviewReducer,
  reviewReplies: reviewReplyReducer,
  tags: tagReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
