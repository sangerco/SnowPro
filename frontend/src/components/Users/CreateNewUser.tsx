import React, { useState } from "react";
import { connect } from "react-redux";
import { sendNewUserData } from "../../oldRedux/actions/userActions";
import { RootState } from "../../oldRedux/store";
import { Button, Input, Form, Message } from "semantic-ui-react";

interface NewUserData {
  username: string;
  password: string;
  first_name: string;
  last_name: string;
  email: string;
}

interface NewUserProps {
  newUser: NewUserData | null;
  loading: boolean;
  error: string;
  sendNewUserData: (
    username: string,
    password: string,
    first_name: string,
    last_name: string,
    email: string
  ) => Promise<void>;
}

const CreateUserForm: React.FC<NewUserProps> = ({
  newUser,
  loading,
  error,
  sendNewUserData,
}) => {
  const [formData, setFormData] = useState<NewUserData>({
    username: "",
    password: "",
    first_name: "",
    last_name: "",
    email: "",
  });

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    await sendNewUserData(
      formData.username,
      formData.password,
      formData.first_name,
      formData.last_name,
      formData.email
    );

    setFormData({
      username: "",
      password: "",
      first_name: "",
      last_name: "",
      email: "",
    });
  };

  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;
    setFormData({
      ...formData,
      [name]: value,
    });
  };

  return (
    <Form onSubmit={handleSubmit} error>
      <Form.Field>
        <label>First Name</label>
        <input
          placeholder="First Name"
          name="first_name"
          value={formData.first_name}
          onChange={handleChange}
        />
      </Form.Field>
      <Form.Field>
        <label>Last Name</label>
        <input
          placeholder="Last Name"
          name="last_name"
          value={formData.last_name}
          onChange={handleChange}
        />
      </Form.Field>
      <Form.Field>
        <label>Username</label>
        <input
          placeholder="username"
          name="username"
          value={formData.username}
          onChange={handleChange}
        />
      </Form.Field>
      <Form.Field>
        <label>Choose Password</label>
        <Input
          type="password"
          name="password"
          value={formData.password}
          onChange={handleChange}
        />
      </Form.Field>
      <Form.Field>
        <label>Email</label>
        <Input
          type="email"
          name="email"
          value={formData.email}
          onChange={handleChange}
        />
      </Form.Field>
      <Button type="submit">Submit</Button>
      {loading && <p>Loading...</p>}
      <Message error header={error} content={error} />
    </Form>
  );
};

const mapStateToProps = (state: RootState) => ({
  newUser: state.newUser.data,
  loading: state.newUser.loading,
  error: state.newUser.error,
});

export default connect(mapStateToProps, { sendNewUserData })(CreateUserForm);
