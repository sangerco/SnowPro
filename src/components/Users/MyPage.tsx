import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  Dimmer,
  Loader,
  Grid,
  Container,
  Header,
  Image,
  Modal,
  Button,
  Icon,
} from "semantic-ui-react";
import { Link } from "react-router-dom";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchOneUser, deleteUser } from "../../redux/slices/userSlice";

const MyPage = () => {
  const auth = useSelector((state: RootState) => state.auth);
  const users = useSelector((state: RootState) => state.users);
  const user = users.user;
  const dispatch = useDispatch<AppDispatch>();
  const username = auth.data?.username;

  const [showDeleteModal, setShowDeleteModal] = useState(false);

  useEffect(() => {
    if (username) {
      dispatch(fetchOneUser(username));
    }
  }, [dispatch, username]);

  const handleShowDeleteModal = () => {
    setShowDeleteModal(true);
  };

  const handleDeleteUser = () => {
    if (username) {
      dispatch(deleteUser(username));
    }
  };

  if (users.loading) {
    return (
      <div>
        <Dimmer active>
          <Loader>Loading...</Loader>
        </Dimmer>
      </div>
    );
  }

  if (users.error) {
    return <div>Error: {users.error}</div>;
  }

  if (user) {
    return (
      <div>
        <Container>
          <Grid relaxed>
            <Grid.Row>
              <Grid.Column width={4}>
                <Image src={user.avatar} size="small" />
              </Grid.Column>
              <Grid.Column width={8}>
                <Header as="h2" textAlign="right">
                  {user.username}
                </Header>
                <Header.Subheader>
                  {user.firstName && user.firstName !== ""
                    ? user.firstName
                    : ""}
                  {user.lastName && user.lastName !== "" ? user.lastName : ""}
                </Header.Subheader>
                <Link to={"madeuplinkfornow"}>
                  <Icon name="edit" style={{ cursor: "pointer" }} />
                </Link>
                <Icon
                  name="trash"
                  style={{ cursor: "pointer" }}
                  onClick={handleShowDeleteModal()}
                />
              </Grid.Column>
              <Grid.Column width={4}></Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={4}></Grid.Column>
              <Grid.Column width={8}>
                <p>{user.bio}</p>
              </Grid.Column>
              <Grid.Column width={4}>
                {/* {user.favMountains.length > 0 ? (
                    <FavMountain userId={user.id} />
                ) : (
                    <p>This user has no favorited mountains yet!</p>
                )} */}
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={2}></Grid.Column>
              <Grid.Column width={12}>
                {/* <ShowPhotos username={username as string} /> */}
              </Grid.Column>
              <Grid.Column width={2}></Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={2}></Grid.Column>
              <Grid.Column width={12}>
                {/* <ShowVideos username={username as string} /> */}
              </Grid.Column>
              <Grid.Column width={2}></Grid.Column>
            </Grid.Row>
          </Grid>
        </Container>
        <Modal open={showDeleteModal} onClose={() => setShowDeleteModal(false)}>
          <Modal.Content>
            Are You Sure You Want To Delete Your Profile?
          </Modal.Content>
          <Modal.Actions>
            <Button negative onClick={handleDeleteUser}>
              Yes
            </Button>
            <Button onClick={() => setShowDeleteModal(false)}>Cancel</Button>
          </Modal.Actions>
        </Modal>
      </div>
    );
  }

  return null;
};

export default MyPage;
