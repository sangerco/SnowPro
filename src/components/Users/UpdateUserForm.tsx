import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { RootState, AppDispatch } from "../../redux/store";
import {
  UserData,
  fetchOneUser,
  updateUser,
} from "../../redux/slices/userSlice";
import {
  Container,
  Form,
  Button,
  Dimmer,
  Loader,
  Header,
} from "semantic-ui-react";

const UpdateUserForm: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const users = useSelector((state: RootState) => state.users);
  const user = users.user;

  useEffect(() => {
    if (user) {
      dispatch(fetchOneUser(user.username));
    }
  }, [dispatch, user]);

  let initialState: UserData = {
    id: "",
    username: "",
    password: "",
    email: "",
    firstName: "",
    lastName: "",
    isAdmin: false,
    avatar: "",
    bio: "",
    vLinks: [],
    pLinks: [],
    favMountains: [],
  };

  useEffect(() => {
    if (user) {
      setFormData({
        id: user.id,
        username: user.username,
        password: user.password || "",
        email: user.email,
        firstName: user.firstName,
        lastName: user.lastName,
        isAdmin: user.isAdmin,
        avatar: user.avatar || "",
        bio: user.bio || "",
        vLinks: user.vLinks || [],
        pLinks: user.pLinks || [],
        favMountains: user.favMountains || [],
      });
      setPhotos(user.pLinks || []);
      setVideos(user.vLinks || []);
    }
  }, [user]);

  const [formData, setFormData] = useState(initialState);
  const [photos, setPhotos] = useState(user?.pLinks);
  const [videos, setVideos] = useState(user?.vLinks);

  if (users.loading) {
    return (
      <div>
        <Dimmer active>
          <Loader>Loading...</Loader>
        </Dimmer>
      </div>
    );
  }

  if (users.error) {
    return (
      <div>
        <Dimmer>
          <p>Error! User data cannot be found! {`${users.error}`}</p>
        </Dimmer>
      </div>
    );
  }

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (formData) {
      dispatch(updateUser(formData));
    }

    setFormData(initialState);
  };

  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handleTextAreaChange: React.ChangeEventHandler<HTMLTextAreaElement> = (
    e
  ) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handlePhotoChange = (index: number, value: string) => {
    if (formData.pLinks) {
      const updatedPhotoLinks = [...formData.pLinks];
      updatedPhotoLinks[index] = value;
      setPhotos(updatedPhotoLinks);
    }
  };

  const handleAddPhotoInput = () => {
    if (formData.pLinks) {
      setPhotos([...formData.pLinks, ""]);
    }
  };

  const handleRemovePhotoInput = (index: number) => {
    if (formData.pLinks) {
      const updatedPhotoLinks = [...formData.pLinks];
      updatedPhotoLinks.splice(index, 1);
      setPhotos(updatedPhotoLinks);
    }
  };

  const handleVideoChange = (index: number, value: string) => {
    if (formData.vLinks) {
      const updatedVideoLinks = [...formData.vLinks];
      updatedVideoLinks[index] = value;
      setPhotos(updatedVideoLinks);
    }
  };

  const handleAddVideoInput = () => {
    if (formData.vLinks) {
      setVideos([...formData.vLinks, ""]);
    }
  };

  const handleRemoveVideoInput = (index: number) => {
    if (formData.vLinks) {
      const updatedVideoLinks = [...formData.vLinks];
      updatedVideoLinks.splice(index, 1);
      setVideos(updatedVideoLinks);
    }
  };

  if (user) {
    return (
      <div>
        <Container>
          <Header as="h2">{`Update ${user.username}'s Profile`}</Header>
          <Form onSubmit={handleSubmit}>
            <Form.Field>
              <label>Username</label>
              <input
                placeholder={formData.username}
                name="username"
                value={formData.username}
                onChange={handleChange}
              />
            </Form.Field>
            <Form.Field>
              <label>Password</label>
              <input
                type="password"
                placeholder={formData.password}
                name="password"
                value={formData.password}
                onChange={handleChange}
              />
            </Form.Field>
            <Form.Field>
              <label>First Name</label>
              <input
                placeholder={formData.firstName}
                name="firstName"
                value={formData.firstName}
                onChange={handleChange}
              />
            </Form.Field>
            <Form.Field>
              <label>Last Name</label>
              <input
                placeholder={formData.lastName}
                name="lastName"
                value={formData.lastName}
                onChange={handleChange}
              />
            </Form.Field>
            <Form.Field>
              <label>Email</label>
              <input
                type="email"
                placeholder={formData.email}
                name="email"
                value={formData.email}
                onChange={handleChange}
              />
            </Form.Field>
            <Form.Field>
              <label>About You...</label>
              <textarea
                placeholder={formData.bio}
                name="bio"
                value={formData.bio}
                onChange={handleTextAreaChange}
              />
            </Form.Field>
            <Form.Field>
              <label>Avatar Link</label>
              <input
                placeholder={formData.avatar}
                name="avatar"
                value={formData.avatar}
                onChange={handleChange}
              />
            </Form.Field>
            {formData.pLinks &&
              formData.pLinks.map((photo, index) => (
                <Form.Field key={index}>
                  <label>Photos</label>
                  <input
                    value={photo}
                    onChange={(e) => handlePhotoChange(index, e.target.value)}
                  />
                  {formData.pLinks && index === formData.pLinks.length - 1 && (
                    <Button icon="plus" onClick={handleAddPhotoInput} />
                  )}
                  {index > 0 && (
                    <Button
                      icon="minus"
                      onClick={() => handleRemovePhotoInput(index)}
                    />
                  )}
                </Form.Field>
              ))}
            {formData.vLinks &&
              formData.vLinks.map((video, index) => (
                <Form.Field key={index}>
                  <label>Videos</label>
                  <input
                    value={video}
                    onChange={(e) => handleVideoChange(index, e.target.value)}
                  />
                  {formData.vLinks && index === formData.vLinks.length - 1 && (
                    <Button icon="plus" onClick={handleAddVideoInput} />
                  )}
                  {index > 0 && (
                    <Button
                      icon="minus"
                      onClick={() => handleRemoveVideoInput(index)}
                    />
                  )}
                </Form.Field>
              ))}
            <Button type="submit">Submit Changes</Button>
          </Form>
        </Container>
      </div>
    );
  }

  return null;
};

export default UpdateUserForm;
