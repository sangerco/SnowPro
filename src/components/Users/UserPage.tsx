import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  Dimmer,
  Loader,
  Grid,
  Container,
  Header,
  Image,
  Modal,
  Button,
  Icon,
  Card,
  Divider,
  Embed,
} from "semantic-ui-react";
import { Link, useParams } from "react-router-dom";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchOneUser, deleteUser } from "../../redux/slices/userSlice";
import Inbox from "../Messages/Inbox";
import FavMountain from "../SkiAreas/FavMountain";

const UserPage: React.FC = () => {
  const auth = useSelector((state: RootState) => state.auth);
  const authId = auth.data?.id;
  const { username } = useParams();
  const users = useSelector((state: RootState) => state.users);
  const user = users.user;
  console.log(user?.favMountains);
  const dispatch = useDispatch<AppDispatch>();

  const [showDeleteModal, setShowDeleteModal] = useState(false);

  useEffect(() => {
    if (username) {
      dispatch(fetchOneUser(username));
    }
  }, [dispatch, username]);

  const handleShowDeleteModal = () => {
    setShowDeleteModal(true);
  };

  const handleDeleteUser = () => {
    if (username) {
      dispatch(deleteUser(username));
    }
  };

  let videoIds = [];

  if (user && user.videos && user.videos.length > 0) {
    for (let i = 0; i < user.videos.length; i++) {
      let splitLink = user.videos[i].split("v=");
      let id = splitLink[1];
      let videoId = id.includes("&") ? id.split("&")[0] : id;
      videoIds.push(videoId);
    }
  }

  if (username && user?.favMountains && user.favMountains.length > 0) {
    console.log("true");
  }

  if (users.loading) {
    return (
      <div>
        <Dimmer active>
          <Loader>Loading...</Loader>
        </Dimmer>
      </div>
    );
  }

  if (users.error) {
    return <div>Error: {users.error}</div>;
  }

  if (user) {
    return (
      <div>
        <Container>
          <Grid relaxed>
            <Grid.Row>
              <Grid.Column width={4}>
                <Image src={user.avatar} size="small" />
              </Grid.Column>
              <Grid.Column width={8}>
                <Header as="h2" textAlign="right">
                  {user.username}
                </Header>
                <Header.Subheader>
                  {user.firstName && user.firstName !== ""
                    ? user.firstName
                    : ""}{" "}
                  {user.lastName && user.lastName !== "" ? user.lastName : ""}
                </Header.Subheader>
                {authId && authId === user.id ? (
                  <div>
                    <Link to={`/users/${user.username}/update`}>
                      <Icon name="edit" style={{ cursor: "pointer" }} />
                    </Link>
                    <Icon
                      name="trash"
                      style={{ cursor: "pointer" }}
                      onClick={handleShowDeleteModal}
                    />
                  </div>
                ) : null}
              </Grid.Column>
              <Grid.Column width={4}></Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={4}></Grid.Column>
              <Grid.Column width={8}>
                <p>{user.bio}</p>
              </Grid.Column>
              <Grid.Column width={4}>
                <Header as="h3">Favorite Mountains</Header>
                <Divider />
                {user.favMountains &&
                username &&
                user.favMountains.length > 0 ? (
                  user.favMountains.map((fm) => (
                    <Card.Group>
                      <FavMountain key={fm} slug={fm} username={username} />
                    </Card.Group>
                  ))
                ) : (
                  <p>This user has no favorited mountains yet!</p>
                )}
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={2}></Grid.Column>
              <Grid.Column width={12}>
                <Header as="h3">
                  <Link to={`/users/${username}/photos`}>
                    All {username}'s Photos
                  </Link>
                </Header>
                {user.photos && user.photos.length > 0 ? (
                  user.photos.map((photo) => <Image size="small" src={photo} />)
                ) : (
                  <p>This user has no photos yet!</p>
                )}
                <Divider />
              </Grid.Column>
              <Grid.Column width={2}></Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column width={2}></Grid.Column>
              <Grid.Column width={12}>
                <Header as="h3">
                  <Link to={`/users/${username}/videos`}>
                    All {username}'s Videos
                  </Link>
                </Header>
                {videoIds && videoIds.length > 0 ? (
                  videoIds.map((video) => (
                    <Embed size="small" id={video} source="youtube" />
                  ))
                ) : (
                  <p>This user has no videos yet!</p>
                )}
                <Divider />
              </Grid.Column>
              <Grid.Column width={2}></Grid.Column>
            </Grid.Row>
          </Grid>
        </Container>
        <Modal open={showDeleteModal} onClose={() => setShowDeleteModal(false)}>
          <Modal.Content>
            Are You Sure You Want To Delete Your Profile?
          </Modal.Content>
          <Modal.Actions>
            <Button negative onClick={handleDeleteUser}>
              Yes
            </Button>
            <Button onClick={() => setShowDeleteModal(false)}>Cancel</Button>
          </Modal.Actions>
        </Modal>
      </div>
    );
  }

  return null;
};

export default UserPage;
