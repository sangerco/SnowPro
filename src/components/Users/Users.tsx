import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { RootState, AppDispatch } from "../../redux/store";
import { fetchUserData, UserData } from "../../redux/slices/userSlice";
import {
  Button,
  Card,
  Image,
  Loader,
  Dimmer,
  Grid,
  Rating,
  Divider,
  Header,
} from "semantic-ui-react";
import { Link } from "react-router-dom";
import { fetchSkiAreas } from "../../redux/slices/skiAreaSlice";
import { fetchAllReviews } from "../../redux/slices/reviewSlice";

const Users = () => {
  const users = useSelector((state: RootState) => state.users);
  const dispatch = useDispatch<AppDispatch>();

  useEffect(() => {
    dispatch(fetchUserData());
  }, [dispatch]);

  useEffect(() => {
    dispatch(fetchSkiAreas());
  }, [dispatch]);

  const skiAreaState = useSelector((state: RootState) => state.skiAreas);
  const skiAreas = skiAreaState.skiAreas;

  useEffect(() => {
    dispatch(fetchAllReviews());
  }, [dispatch]);

  const reviewState = useSelector((state: RootState) => state.reviews);
  const reviews = reviewState.reviews;

  const formatDate = (date: Date) => {
    const newDate = new Date(date);
    const options: Intl.DateTimeFormatOptions = {
      year: "numeric",
      month: "long",
      day: "numeric",
    };
    return newDate.toLocaleDateString("en-US", options);
  };

  if (users.loading) {
    <div>
      <Dimmer active>
        <Loader>Loading...</Loader>
      </Dimmer>
    </div>;
  }

  if (users.error) {
    <div>
      <Dimmer active>
        <div>{`Error! Users cannot be loaded! ${users.error}`}</div>
      </Dimmer>
    </div>;
  }

  if (users.users) {
    return (
      <div>
        <Grid>
          <Grid.Row>
            <Grid.Column widescreen={4}>
              {skiAreaState.loading ? (
                <Dimmer active>
                  <Loader>Loading...</Loader>
                </Dimmer>
              ) : skiAreas && skiAreas.length > 0 ? (
                <Card id="ski-area-card">
                  {" "}
                  {/* max height overflow auto */}
                  {skiAreas.map((sa) => (
                    <Card.Content key={sa.slug}>
                      <Link to={`/ski-areas/${sa.slug}`}>{sa.name}</Link>
                    </Card.Content>
                  ))}
                </Card>
              ) : skiAreaState.error ? (
                <Dimmer active>
                  <Header as="h1">
                    Error! Ski Area Data cannot be retrieved!{" "}
                    {skiAreaState.error}
                  </Header>
                </Dimmer>
              ) : null}
            </Grid.Column>
            <Grid.Column width={8}>
              {users.users.map((user: UserData) => (
                <Card.Group key={user.id}>
                  <Card>
                    <Card.Content>
                      {user.avatar ? (
                        <Image floated="right" wrapped src={user.avatar} />
                      ) : null}
                      <Card.Header>{user.username}</Card.Header>
                      <Card.Meta>
                        {user.firstName} {user.lastName}
                      </Card.Meta>
                    </Card.Content>
                    <Card.Content extra>
                      <Button as={Link} to={`/users/${user.username}`}>
                        {`${user.username}'s Profile`}
                      </Button>
                    </Card.Content>
                  </Card>
                </Card.Group>
              ))}
            </Grid.Column>
            <Grid.Column width={4}>
              {reviewState.loading ? (
                <Dimmer active>
                  <Loader>Loading...</Loader>
                </Dimmer>
              ) : reviewState.error ? (
                <Dimmer active>
                  <Header as="h1">
                    Error! Review Data cannot be retrieved! {reviewState.error}
                  </Header>
                </Dimmer>
              ) : reviews && reviews.length > 0 ? (
                <>
                  {reviews.map((review) => (
                    <Card>
                      {" "}
                      {/* paginated */}
                      <Card.Content key={review.id} id="review-card">
                        <Card.Header>
                          <Link to={`/ski-areas/reviews/${review.id}`}>
                            {review.header}
                          </Link>{" "}
                          <Divider />
                          {review.skiAreaName}
                        </Card.Header>
                        <Card.Description>
                          By {review.username}
                        </Card.Description>
                        <Card.Meta>
                          <Rating
                            icon="star"
                            defaultRating={review.stars}
                            maxRating={5}
                            disabled
                          />
                        </Card.Meta>
                        {formatDate(review.createdAt)}
                      </Card.Content>
                    </Card>
                  ))}
                </>
              ) : null}
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </div>
    );
  }

  return null;
};

export default Users;
