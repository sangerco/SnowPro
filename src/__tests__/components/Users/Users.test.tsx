/* eslint-disable testing-library/no-node-access */
import React from "react";
import Users from "../../../components/Users/Users";
import { render, fireEvent, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../../redux/store";
import { BrowserRouter } from "react-router-dom";

describe("test Users", () => {
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Users />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Users />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
