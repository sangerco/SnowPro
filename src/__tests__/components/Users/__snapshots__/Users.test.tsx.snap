// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Users should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="ui grid"
    >
      <div
        class="row"
      >
        <div
          class="four wide widescreen column"
        >
          <div
            class="ui segment"
            style="overflow: auto; max-height: 75vh;"
          >
            <div
              class="ui card"
              id="ski-area-card"
            >
               
              <div
                class="content"
              >
                <a
                  href="/ski-areas/ski-area-1"
                >
                  Ski Area 1
                </a>
              </div>
              <div
                class="content"
              >
                <a
                  href="/ski-areas/ski-area-1"
                >
                  Ski Area 1
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="eight wide column"
        >
          <div
            class="ui cards"
          >
            <div
              class="ui card"
            >
              <div
                class="content"
                data-testid="user"
              >
                <div
                  class="ui right floated image"
                >
                  <img
                    src="https://testdata.com/testavatar.jpg"
                  />
                </div>
                <div
                  class="header"
                  data-testid="user-header"
                >
                  testuser
                </div>
                <div
                  class="meta"
                >
                  Test User
                </div>
              </div>
              <div
                class="extra content"
              >
                <a
                  class="ui blue button"
                  href="/users/testuser"
                  role="button"
                >
                  testuser's Profile
                </a>
              </div>
            </div>
            <div
              class="ui card"
            >
              <div
                class="content"
                data-testid="user"
              >
                <div
                  class="ui right floated image"
                >
                  <img
                    src="https://testdata.com/testavatar2.jpg"
                  />
                </div>
                <div
                  class="header"
                  data-testid="user-header"
                >
                  testuser2
                </div>
                <div
                  class="meta"
                >
                  Test User2
                </div>
              </div>
              <div
                class="extra content"
              >
                <a
                  class="ui blue button"
                  href="/users/testuser2"
                  role="button"
                >
                  testuser2's Profile
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="four wide column"
        >
          <div
            class="ui card"
          >
             
            <div
              class="content"
              id="review-card"
            >
              <div
                class="header"
              >
                <a
                  href="/ski-areas/reviews/review-1"
                >
                  test review
                </a>
                 
                <div
                  class="ui divider"
                />
                Ski Area 1
              </div>
              <div
                class="description"
              >
                By testuser
              </div>
              <div
                class="meta"
              >
                <div
                  class="ui star disabled rating"
                  role="radiogroup"
                  tabindex="0"
                >
                  <i
                    aria-checked="false"
                    aria-posinset="1"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="2"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="3"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="4"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="true"
                    aria-posinset="5"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                </div>
              </div>
              Invalid Date
            </div>
          </div>
          <div
            class="ui card"
          >
             
            <div
              class="content"
              id="review-card"
            >
              <div
                class="header"
              >
                <a
                  href="/ski-areas/reviews/review-2"
                >
                  test review
                </a>
                 
                <div
                  class="ui divider"
                />
                Ski Area 2
              </div>
              <div
                class="description"
              >
                By testuser2
              </div>
              <div
                class="meta"
              >
                <div
                  class="ui star disabled rating"
                  role="radiogroup"
                  tabindex="0"
                >
                  <i
                    aria-checked="false"
                    aria-posinset="1"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="true"
                    aria-posinset="2"
                    aria-setsize="5"
                    class="active icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="3"
                    aria-setsize="5"
                    class="icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="4"
                    aria-setsize="5"
                    class="icon"
                    role="radio"
                    tabindex="-1"
                  />
                  <i
                    aria-checked="false"
                    aria-posinset="5"
                    aria-setsize="5"
                    class="icon"
                    role="radio"
                    tabindex="-1"
                  />
                </div>
              </div>
              Invalid Date
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
